File: app/src/main/java/com/bcg/gpscompass/ui/screen/compass/CompassFragment.java
package com.bcg.gpscompass.ui.screen.compass;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.graphics.SurfaceTexture;
import android.graphics.Typeface;
import android.hardware.Camera;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.location.Geocoder;
import android.location.Location;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.TranslateAnimation;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.bcg.gpscompass.R;
import com.bcg.gpscompass.ui.base.BaseFragment;
import com.bcg.gpscompass.ui.view.CompassImageView;
import com.bcg.gpscompass.utils.Constants;
import com.bcg.gpscompass.utils.gps.GpsLocationService;
import com.bcg.gpscompass.utils.gps.GpsUtil;
import com.bcg.gpscompass.utils.sensor.SensorManagerCompass;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.LocationSettingsStatusCodes;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class CompassFragment extends BaseFragment<CompassPresenter> implements SensorEventListener, CompassListener, View.OnClickListener {

    private SensorManagerCompass mSensorManagerCompass;
    private CompassImageView mCustomImageCompassView;
    private TextView mTxtAddress;
    private TextView mTvDegreesDirection;
    private ImageButton mBtnMap;
    private ImageButton mBtnLocation;
    private ImageButton mBtnWeather;
    private Typeface mTypeface;
    private int mCurrentMagnetic = 0;
    private float[] valuesAccel = new float[]{0f, 0f, 9.8f};
    private float[] valuesMagnetic = new float[]{0.5f, 0f, 0f};
    private float mAzimuth;

    private FusedLocationProviderClient fusedLocationProviderClient;
    private Location mLastLocation;
    private LocationRequest mLocationRequest;
    private AddressResultReceiver addressResultReceiver;
    private String mLocationOutput;
    private String mFullLocation;
    private static final int CURRENT_REQUEST = 112;
    private float[] mGravity;
    private double currentInclinationX;
    private double currentInclinationY;
    private double startInclinationX;
    private double startInclinationY;
    private ImageView ivBallCompass;
    private float AA = 0.3F;

    public CompassFragment() {
    }

    public static CompassFragment newInstance() {
        return new CompassFragment();
    }

    @Override
    protected CompassPresenter createPresenter() {
        return new CompassPresenter(this);
    }

    @SuppressLint("MissingPermission")
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        mTypeface = ResourcesCompat.getFont(getActivity(), R.font.font_roboto_bold);
        mSensorManagerCompass = new SensorManagerCompass(requireActivity());
        addressResultReceiver = new AddressResultReceiver(new Handler());
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(Objects.requireNonNull(getActivity()));
        mLocationOutput = "";
        mFullLocation = "";
        createLocationRequest();
        locationSetting();
    }

    private void createLocationRequest() {
        mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(10000);
        mLocationRequest.setFastestInterval(5000);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    }

    @SuppressLint("MissingPermission")
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_compass, container, false);
        //RateUsDialogApp.showDialogRateApp(getActivity());
        mBtnMap = (ImageButton) view.findViewById(R.id.btn_map_compass);
        mBtnLocation = (ImageButton) view.findViewById(R.id.btn_location_compass);
        mCustomImageCompassView = (CompassImageView) view.findViewById(R.id.iv_compass);
        mTxtAddress = (TextView) view.findViewById(R.id.tv_address);
        //mTvDegreesDirection = (TextView) view.findViewById(R.id.tv_degrees_direction_home);
//        mTvCity.setTypeface(mTypeface);
//        mTvCity.setEllipsize(TextUtils.TruncateAt.MARQUEE);
//        mTvCity.setSelected(true);
//        mTvCity.setText("Wait...");
//        mTvDegreesDirection.setTypeface(mTypeface);
//        nativeExpressAdView = view.findViewById(R.id.ads_banner_home);

//        rlLimited = view.findViewById(R.id.rl_limit);
//        ivBallCompass = view.findViewById(R.id.iv_ball);

        mBtnWeather = view.findViewById(R.id.btn_weather_compass);
        mBtnWeather.setOnClickListener(this);
        mBtnLocation.setOnClickListener(this);
        mBtnMap.setOnClickListener(this);

//        admobBannerUnit();
        getLastLocation();
        return view;
    }

    LocationCallback mLocationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            if (locationResult == null) {
                return;
            }
            for (Location location : locationResult.getLocations()) {
                mPresenter.setTextLocation(location.getLatitude(), location.getLongitude());
                mLastLocation = location;
                getLocation();
            }
        }
    };

    public static void xyzAnim(View paramView, double[] paramArrayOfDouble1, double[] paramArrayOfDouble2, int paramInt) {
        try {
            TranslateAnimation localTranslateAnimation = new TranslateAnimation((float) paramArrayOfDouble1[0], (float) paramArrayOfDouble2[0], (float) paramArrayOfDouble1[1], (float) paramArrayOfDouble2[1]);
            localTranslateAnimation.setDuration(paramInt);
            localTranslateAnimation.setFillEnabled(true);
            localTranslateAnimation.setFillAfter(true);
            paramView.startAnimation(localTranslateAnimation);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected float[] xyzBall(float[] mFloat1, float[] mFloat2, float currentFloat) {
        if (mFloat2 == null) {
            return mFloat1;
        }
        for (int i = 0; i < mFloat1.length; i++) {
            mFloat2[i] += currentFloat * (mFloat1[i] - mFloat2[i]);
        }
        return mFloat2;
    }

    private void tiltDevice() {
//        double dbMath = Math.sqrt(mGravity[0] * mGravity[0] + mGravity[1] * mGravity[1] + mGravity[2] * mGravity[2]);
//        double[] dbListArray = new double[3];
//        dbListArray[0] = (mGravity[0] / dbMath);
//        dbListArray[1] = (mGravity[1] / dbMath);
//        dbListArray[2] = (mGravity[2] / dbMath);
//        currentInclinationX = (-Math.toDegrees(Math.asin(dbListArray[0])));
//        currentInclinationY = Math.toDegrees(Math.asin(dbListArray[1]));
//        double dbLayout = rlLimited.getWidth() / 2 - ivBallCompass.getWidth() / 2;
//        currentInclinationX = (dbLayout * -dbListArray[0]);
//        currentInclinationY = (dbLayout * dbListArray[1]);
    }

//    private void admobBannerUnit() {
//        final AdView mAdView = new AdView(Objects.requireNonNull(getContext()));
//        final AdRequest adRequest = new AdRequest.Builder().build();
//        mAdView.setAdSize(AdSize.BANNER);
//        mAdView.setAdUnitId(getString(R.string.banner_id_1));
////        nativeExpressAdView.addView(mAdView);
//        mAdView.loadAd(adRequest);
//        mAdView.setAdListener(new AdListener() {
//            @Override
//            public void onAdClosed() {
//                super.onAdClosed();
//            }
//
//            @Override
//            public void onAdOpened() {
//                super.onAdOpened();
//            }
//
//            @Override
//            public void onAdLeftApplication() {
//                super.onAdLeftApplication();
//            }
//
//            @Override
//            public void onAdFailedToLoad(int i) {
//                super.onAdFailedToLoad(i);
//            }
//
//            @Override
//            public void onAdLoaded() {
//                super.onAdLoaded();
//            }
//        });
//    }

    @SuppressLint("MissingPermission")
    private void getLastLocation() {
        fusedLocationProviderClient.getLastLocation()
                .addOnSuccessListener(Objects.requireNonNull(getActivity()), new OnSuccessListener<Location>() {
                    @Override
                    public void onSuccess(Location location) {
                        if (location != null) {
                            mLastLocation = location;
                            mPresenter.setTextLocation(location.getLatitude(), location.getLongitude());
                            getLocation();
                        } else {
                        }
                    }
                });
    }

    private void getLocation() {
        if (!Geocoder.isPresent()) {
            Toast.makeText(getActivity(),
                    "Can't find current address, ",
                    Toast.LENGTH_SHORT).show();
            return;
        }
        startIntentService();
    }

    protected void startIntentService() {
        Intent intent = new Intent(getActivity(), GpsLocationService.class);
        intent.putExtra(Constants.RECEIVER_EXTRA, addressResultReceiver);
        intent.putExtra(Constants.LOCATION_EXTRA, mLastLocation);
        getActivity().startService(intent);
    }

    @Override
    public void onResume() {
        super.onResume();
        mSensorManagerCompass.registerAccelerometerListener(this);
        mSensorManagerCompass.registerMagneListener(this);
        mSensorManagerCompass.registerOriListener(this);
        startLocationUpdates();
    }

    @SuppressLint("MissingPermission")
    private void startLocationUpdates() {
        fusedLocationProviderClient.requestLocationUpdates(mLocationRequest, mLocationCallback, null);
    }

    @Override
    public void onPause() {
        super.onPause();
        mSensorManagerCompass.unregisterListener(this);
        stopUpdateLocation();
    }

    private void stopUpdateLocation() {
        fusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
    }

    @Override
    public void onSensorChanged(SensorEvent sensorEvent) {
        mPresenter.changeDirection(sensorEvent);
        if (sensorEvent.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            mGravity = xyzBall(sensorEvent.values.clone(), mGravity, AA);
            tiltDevice();
            double[] a1 = new double[2];
            a1[0] = startInclinationX;
            a1[1] = startInclinationY;
            double[] a2 = new double[2];
            a2[0] = (-currentInclinationX);
            a2[1] = (-currentInclinationY);
            //xyzAnim(ivBallCompass, a1, a2, 210);
            startInclinationX = (-currentInclinationX);
            startInclinationY = (-currentInclinationY);
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int i) {
        if (sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD) {
            switch (i) {
                case SensorManager.SENSOR_STATUS_ACCURACY_LOW:
                    mPresenter.calibrateCompass(SensorManager.SENSOR_STATUS_ACCURACY_LOW);
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_LOW;
                    break;
                case SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM:
                    mPresenter.calibrateCompass(SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM);
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM;
                    break;
                case SensorManager.SENSOR_STATUS_ACCURACY_HIGH:
                    mPresenter.calibrateCompass(SensorManager.SENSOR_STATUS_ACCURACY_HIGH);
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_HIGH;
                    break;
            }
        }
    }

    @Override
    public void setChangeDirectionCompass(SensorEvent sensorEvent) {
        if (sensorEvent.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            valuesAccel = GpsUtil.lowPass(sensorEvent.values, valuesAccel);
            mSensorManagerCompass.setGravity(valuesAccel);
        } else if (sensorEvent.sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD) {
            valuesMagnetic = GpsUtil.lowPass(sensorEvent.values,
                    valuesMagnetic);
            mSensorManagerCompass.setGeoMagnetic(valuesMagnetic);
        }
        mSensorManagerCompass.loadAzimuth();
        float newAzimuth = mSensorManagerCompass.getAzimuthSensor();
        if (mAzimuth != newAzimuth) {
            mAzimuth = newAzimuth;
            mCustomImageCompassView.setDegress(-mAzimuth);
            mCustomImageCompassView.invalidate();
            int degrees = Math.round(mAzimuth);
            String direction = GpsUtil.displayDirection(mAzimuth);
            //mTvDegreesDirection.setText(getString(R.string.text_direction, degrees, direction));
        }
    }

    @Override
    public void setTextForLocation(double lat, double lon) {
        String latitude = GpsUtil.decimalToDMS(lat) + GpsUtil.getSymbol(lat, true);
        String longitude = GpsUtil.decimalToDMS(lon) + GpsUtil.getSymbol(lon, false);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {

            case R.id.btn_map_compass:
                mPresenter.openViewGoogleMaps();
                break;
            case R.id.btn_location_compass:
                mPresenter.openLocation();
                break;
            case R.id.btn_weather_compass:
//                Intent mWeather = new Intent(getActivity(), MyWeatherActivity.class);
//                startActivity(mWeather);
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == CURRENT_REQUEST) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                getCamera();
            } else {
                Toast.makeText(getContext(), "The app was not allowed to access camera", Toast.LENGTH_LONG).show();
            }
        }
    }

    @SuppressLint("LongLogTag")
    private void getCamera() {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onStart() {
        super.onStart();
        getCamera();
    }

    @Override
    public void onStop() {
        super.onStop();
    }

    @Override
    public void showViewGoogleMapsCompass() {
//        Intent intent = new Intent(getActivity(), GoogleMapsActivity.class);
//        startActivity(intent);
    }

    @Override
    public void showLocationCompass() {
//        Intent intent = new Intent(getActivity(), LocationActivity.class);
//        intent.putExtra(AppConstantUtils.LOCATION_EXTRA, mLastLocation);
//        intent.putExtra(AppConstantUtils.LOCATION_STRING_EXTRA, mFullLocation);
//        startActivity(intent);
    }

    @Override
    public void showIconLocation() {
        if (mBtnLocation != null) {
            mBtnLocation.setVisibility(View.VISIBLE);
            mBtnMap.setVisibility(View.VISIBLE);
        } else {
            mBtnLocation.setVisibility(View.INVISIBLE);
            mBtnMap.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void getCalibrate(int calibrate) {
//        if (1 == calibrate || 2 == calibrate) {
//            mBtnWarningApp.setColorFilter(ContextCompat.getColor(Objects.requireNonNull(getActivity()), R.color.color_warning));
//        } else {
//            mBtnWarningApp.setColorFilter(ContextCompat.getColor(Objects.requireNonNull(getActivity()), R.color.color_white));
//        }
    }

    @Override
    public void showWarningCompass() {
//        Intent intent = new Intent(getActivity(), HowToUseActivity.class);
//        intent.putExtra(AppConstantUtils.CALIBRATE_MAGENTIC_COMPASS, mCurrentMagnetic);
//        startActivity(intent);
    }


    class AddressResultReceiver extends ResultReceiver {

        public AddressResultReceiver(Handler handler) {
            super(handler);
        }

        @Override
        protected void onReceiveResult(int resultCode, Bundle resultData) {
            if (resultData == null) {
                return;
            }
            mLocationOutput = resultData.getString(Constants.RESULT_KEY_ADDRESS);
            mFullLocation = resultData.getString(Constants.RESULT_KEY_ADDRESS_FULL);

            displayAddressOutput(mFullLocation);
            if (resultCode == Constants.SUCCESS_RESULT) {
                mPresenter.showIconLocation();
            }
        }
    }

    private void displayAddressOutput(String currentAddress) {
        mTxtAddress.setText(currentAddress);
    }

    private void locationSetting() {
        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder().addLocationRequest(mLocationRequest);
        Task<LocationSettingsResponse> result = LocationServices.getSettingsClient(getActivity()).checkLocationSettings(builder.build());
        result.addOnCompleteListener(new OnCompleteListener<LocationSettingsResponse>() {
            @Override
            public void onComplete(@NonNull Task<LocationSettingsResponse> task) {
                try {
                    LocationSettingsResponse response = task.getResult(ApiException.class);
                } catch (ApiException exception) {
                    switch (exception.getStatusCode()) {
                        case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
                            try {
                                ResolvableApiException resolvable = (ResolvableApiException) exception;
                                resolvable.startResolutionForResult(getActivity(), Constants.REQUEST_SETTINGS);
                            } catch (IntentSender.SendIntentException | ClassCastException e) {
                                e.printStackTrace();
                            }
                            break;
                        case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
                            break;
                    }
                }
            }
        });
    }
}

File: app/src/main/java/com/bcg/gpscompass/ui/screen/compass/CompassFragment.kt
package com.bcg.gpscompass.ui.screen.compass

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.graphics.Typeface
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.location.Location
import android.os.Bundle
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.TranslateAnimation
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.bcg.gpscompass.MainActivity
import com.bcg.gpscompass.R
import com.bcg.gpscompass.repository.GpsCompassRepositoryImpl
import com.bcg.gpscompass.repository.model.Geocoding
import com.bcg.gpscompass.repository.remote.ApiState
import com.bcg.gpscompass.ui.screen.AppViewModelFactory
import com.bcg.gpscompass.ui.screen.location.LocationFragment
import com.bcg.gpscompass.ui.screen.map.MapFragment
import com.bcg.gpscompass.ui.screen.payment.PaymentFragment
import com.bcg.gpscompass.ui.screen.weather.WeatherFragment
import com.bcg.gpscompass.ui.view.CompassImageView
import com.bcg.gpscompass.utils.gps.GpsUtil
import com.bcg.gpscompass.utils.location.LocationListenerCallback
import com.bcg.gpscompass.utils.location.LocationListenerCallback.LocationUpdateListener
import com.bcg.gpscompass.utils.sensor.SensorManagerCompass
import com.mapbox.android.core.location.LocationEngine
import com.mapbox.android.core.location.LocationEngineProvider
import com.mapbox.android.core.location.LocationEngineRequest
import kotlinx.coroutines.launch
import kotlin.math.asin
import kotlin.math.sqrt

class CompassFragment : Fragment(), SensorEventListener,
    LocationUpdateListener, View.OnClickListener {
    var DEFAULT_INTERVAL_IN_MILLISECONDS = 1000L
    private var mViewModel: CompassViewModel? = null
    private var mSensorManagerCompass: SensorManagerCompass? = null
    private var mCustomImageCompassView: CompassImageView? = null
    private var mTxtAddress: TextView? = null
    private var mBtnMap: ImageButton? = null
    private var mBtnLocation: ImageButton? = null
    private var mBtnWeather: ImageButton? = null
    private var mInAppPurcharse: AppCompatImageView? = null
    lateinit var mEquilibration: RelativeLayout
    lateinit var mIvBall: ImageView
    private val mTypeface: Typeface? = null
    private var mCurrentMagnetic = 0
    private var valuesAccel = floatArrayOf(0f, 0f, 9.8f)
    private var valuesMagnetic = floatArrayOf(0.5f, 0f, 0f)
    private var mAzimuth = 0f
    private var mGravity: FloatArray = FloatArray(3)
    private var currentInclinationX = 0.0f
    private var currentInclinationY = 0.0f
    private var startInclinationX = 0.0f
    private var startInclinationY = 0.0f
    private val AA = 0.3f
    var DEFAULT_MAX_WAIT_TIME = DEFAULT_INTERVAL_IN_MILLISECONDS * 5
    private val requestPermissionLauncher =
        registerForActivityResult<String, Boolean>(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
            if (isGranted) {
                lastLocation
            } else {
                // Explain to the user that the feature is unavailable because the
                // feature requires a permission that the user has denied. At the
                // same time, respect the user's decision. Don't link to system
                // settings in an effort to convince the user to change their
                // decision.
            }
        }
    private var callback: LocationListenerCallback? = null
    private var locationEngine: LocationEngine? = null
    private var currentAddress: String? = null
    private var shortAddress: String? = null
    private var latitude = 0.0
    private var longitude = 0.0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //        mTypeface = ResourcesCompat.getFont(getActivity(), R.font.font_roboto_bold);
        mSensorManagerCompass = SensorManagerCompass(requireActivity())
        mViewModel = ViewModelProvider(this, AppViewModelFactory(GpsCompassRepositoryImpl())).get(
            CompassViewModel::class.java
        )
        callback = LocationListenerCallback(requireActivity(), this)
    }

    private fun xyzBall(
        mFloat1: FloatArray,
        mFloat2: FloatArray,
        currentFloat: Float
    ): FloatArray {
        if (mFloat2 == null) {
            return mFloat1
        }
        for (i in mFloat1.indices) {
            mFloat2[i] += currentFloat * (mFloat1[i] - mFloat2[i])
        }
        return mFloat2
    }

    private fun tiltDevice() {
        val dbMath = sqrt(
            mGravity[0] * mGravity[0] + mGravity[1] * mGravity[1] + mGravity[2] * mGravity[2]
        )
        val dbListArray = FloatArray(3)
        dbListArray[0] = mGravity[0] / dbMath
        dbListArray[1] = mGravity[1] / dbMath
        dbListArray[2] = mGravity[2] / dbMath
        currentInclinationX = -Math.toDegrees(asin(dbListArray[0]).toDouble()).toFloat()
        currentInclinationY = Math.toDegrees(asin(dbListArray[1]).toDouble()).toFloat()
        val dbLayout: Int = mEquilibration.width / 2 - mIvBall.width / 2
        currentInclinationX = dbLayout * -dbListArray[0]
        currentInclinationY = dbLayout * dbListArray[1]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_compass, container, false)
        mBtnMap = view.findViewById<View>(R.id.btn_map_compass) as ImageButton
        mBtnLocation = view.findViewById<View>(R.id.btn_location_compass) as ImageButton
        mCustomImageCompassView = view.findViewById<View>(R.id.iv_compass) as CompassImageView
        mTxtAddress = view.findViewById<View>(R.id.tv_address) as TextView
        mEquilibration = view.findViewById(R.id.rl_equilibration);
        mIvBall = view.findViewById(R.id.iv_ball);
        mBtnWeather = view.findViewById(R.id.btn_weather_compass)
        mInAppPurcharse = view.findViewById(R.id.iv_inapp_purchase)
        mBtnWeather!!.setOnClickListener(this)
        mBtnLocation!!.setOnClickListener(this)
        mBtnMap!!.setOnClickListener(this)
        mInAppPurcharse!!.setOnClickListener(this)
        if (ContextCompat.checkSelfPermission(
                requireActivity(), Manifest.permission.ACCESS_FINE_LOCATION
            ) ==
            PackageManager.PERMISSION_GRANTED
        ) {
            lastLocation
        } else if (shouldShowRequestPermissionRationale("ABC")) {
//            showInContextUI(...);
        } else {
            requestPermissionLauncher.launch(
                Manifest.permission.ACCESS_FINE_LOCATION
            )
        }
        showAddress()
        return view
    }

    private fun showAddress() {
        lifecycleScope.launch {
            mViewModel?.geoCodings?.collect {
                when (it) {
                    is ApiState.Loading -> {
                        mTxtAddress!!.text = "Waiting..."
                    }
                    is ApiState.Failure -> {
                        it.e.printStackTrace()
                        mTxtAddress!!.text = "Not found address"
                    }
                    is ApiState.Success -> {
                        val address = it.data as Geocoding
                        val neighbourhood = if (address.address?.neighbourhood.isNullOrEmpty()) "" else "${address.address?.neighbourhood},"
                        val city = address.address?.city
                        val road = if (address.address?.road.isNullOrEmpty()) "" else "${address.address?.road},"
                        val suburb = if (address.address?.suburb.isNullOrEmpty()) "" else "${address.address?.suburb},"
                        shortAddress = "$road $neighbourhood $suburb $city"
                        shortAddress.also { mTxtAddress!!.text = it }
                        currentAddress = address.displayName
                    }
                    is ApiState.Empty -> {
                        println("Empty...")
                    }
                }

            }
        }
    }

    @get:SuppressLint("MissingPermission")
    private val lastLocation: Unit
        private get() {
            locationEngine = LocationEngineProvider.getBestLocationEngine(requireActivity())
            val request = LocationEngineRequest.Builder(DEFAULT_INTERVAL_IN_MILLISECONDS)
                .setPriority(LocationEngineRequest.PRIORITY_NO_POWER)
                .setMaxWaitTime(DEFAULT_MAX_WAIT_TIME)
                .build()
            locationEngine!!.requestLocationUpdates(request, callback!!, Looper.getMainLooper())
            locationEngine!!.getLastLocation(callback!!)
        }

    override fun onResume() {
        super.onResume()
        mSensorManagerCompass!!.registerAccelerometerListener(this)
        mSensorManagerCompass!!.registerMagneListener(this)
        mSensorManagerCompass!!.registerOriListener(this)
    }

    override fun onPause() {
        super.onPause()
        mSensorManagerCompass!!.unregisterListener(this)
    }

    override fun onSensorChanged(sensorEvent: SensorEvent) {
        setChangeDirectionCompass(sensorEvent)
        if (sensorEvent.sensor.type == Sensor.TYPE_ACCELEROMETER) {
            mGravity = xyzBall(sensorEvent.values.clone(), mGravity, AA)
            tiltDevice()
            val a1 = FloatArray(2)
            a1[0] = startInclinationX
            a1[1] = startInclinationY
            val a2 = FloatArray(2)
            a2[0] = -currentInclinationX
            a2[1] = -currentInclinationY
            xyzAnim(mIvBall, a1, a2, 210);
            startInclinationX = -currentInclinationX
            startInclinationY = -currentInclinationY
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, i: Int) {
        if (sensor.type == Sensor.TYPE_MAGNETIC_FIELD) {
            when (i) {
                SensorManager.SENSOR_STATUS_ACCURACY_LOW -> {
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_LOW
                }
                SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM -> {
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM
                }
                SensorManager.SENSOR_STATUS_ACCURACY_HIGH -> {
                    mCurrentMagnetic = SensorManager.SENSOR_STATUS_ACCURACY_HIGH
                }
            }
        }
    }

    private fun setChangeDirectionCompass(sensorEvent: SensorEvent) {
        if (sensorEvent.sensor.type == Sensor.TYPE_ACCELEROMETER) {
            valuesAccel = GpsUtil.lowPass(sensorEvent.values, valuesAccel)
            mSensorManagerCompass!!.gravity = valuesAccel
        } else if (sensorEvent.sensor.type == Sensor.TYPE_MAGNETIC_FIELD) {
            valuesMagnetic = GpsUtil.lowPass(
                sensorEvent.values,
                valuesMagnetic
            )
            mSensorManagerCompass!!.setGeoMagnetic(valuesMagnetic)
        }
        mSensorManagerCompass!!.loadAzimuth()
        val newAzimuth = mSensorManagerCompass!!.azimuthSensor
        if (mAzimuth != newAzimuth) {
            mAzimuth = newAzimuth
            mCustomImageCompassView!!.setDegress(-mAzimuth)
            mCustomImageCompassView!!.invalidate()
//            val degrees = Math.round(mAzimuth)
//            val direction = GpsUtil.displayDirection(mAzimuth)
            //mTvDegreesDirection.setText(getString(R.string.text_direction, degrees, direction));
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btn_map_compass ->{
                val mapFragment = MapFragment.newInstance(latitude, longitude)
                (requireActivity() as MainActivity).addFragment(mapFragment)
            }
            R.id.btn_location_compass -> {
                val fragment = LocationFragment.newInstance(currentAddress, latitude, longitude)
                (requireActivity() as MainActivity).addFragment(fragment)
            }
            R.id.btn_weather_compass -> {
                val fragment = WeatherFragment.newInstance(shortAddress, latitude, longitude)
                (requireActivity() as MainActivity).addFragment(fragment)
            }
            R.id.iv_inapp_purchase -> {
                val fragment = PaymentFragment.newInstance()
                (requireActivity() as MainActivity).addFragment(fragment)
            }
        }
    }

    override fun onStop() {
        super.onStop()
        if (locationEngine != null) {
            locationEngine!!.removeLocationUpdates(callback!!)
        }
    }

    override fun updateLocation(location: Location) {
        latitude = location.latitude
        longitude = location.longitude
        mViewModel!!.getFlowerList(latitude, longitude)
    }

    companion object {

        @JvmStatic
        fun newInstance(): CompassFragment {
            return CompassFragment()
        }

        fun xyzAnim(
            paramView: ImageView,
            paramArrayOfDouble1: FloatArray,
            paramArrayOfDouble2: FloatArray,
            paramInt: Int
        ) {
            try {
                val localTranslateAnimation = TranslateAnimation(
                    paramArrayOfDouble1[0],
                    paramArrayOfDouble2[0],
                    paramArrayOfDouble1[1],
                    paramArrayOfDouble2[1]
                )
                localTranslateAnimation.duration = paramInt.toLong()
                localTranslateAnimation.isFillEnabled = true
                localTranslateAnimation.fillAfter = true
                paramView.startAnimation(localTranslateAnimation)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}